src/plugins/data/public/antlr/shared/utils.ts

```diff
@@ -244,3 +244,101 @@ export const parseQuery = <
 
   return enrichAutocompleteResult(result, rules, tokenStream, cursorTokenIndex, cursor, query);
 };
+
+export const parseQuery = <
+  A extends AutocompleteResultBase,
+  L extends LexerType,
+  P extends ParserType
+>({
+  Lexer,
+  Parser,
+  tokenDictionary,
+  ignoredTokens,
+  rulesToVisit,
+  getParseTree,
+  enrichAutocompleteResult,
+  query,
+  cursor,
+  context,
+}: ParsingSubject<A, L, P>): AutocompleteResultBase => {
+  const result = singleParseQuery({
+    Lexer,
+    Parser,
+    tokenDictionary,
+    ignoredTokens,
+    rulesToVisit,
+    getParseTree,
+    enrichAutocompleteResult,
+    query,
+    cursor,
+    context,
+  });
+
+  if (!result.rerunWithoutRules) {
+    return result;
+  }
+
+  // go through each rule in list and run a singleParseQuery without it, combining results
+  result.rerunWithoutRules.forEach((rule) => {
+    const modifiedRulesToVisit = new Set(rulesToVisit);
+    modifiedRulesToVisit.delete(rule);
+
+    const nextResult = singleParseQuery({
+      Lexer,
+      Parser,
+      tokenDictionary,
+      ignoredTokens,
+      rulesToVisit: modifiedRulesToVisit,
+      getParseTree,
+      enrichAutocompleteResult,
+      query,
+      cursor,
+      context,
+    });
+
+    // combine result and nextResult
```

Comment:
nit. good to have comments to describe the intent and any potential edge cases as I feel that the for and switch statement handles various data types and use cases, but it might not be immediately clear why certain cases (e.g., boolean or undefined) are treated a specific way.

Reply:
I'll add a big comment outlining the details on combining results